File: scripts/assets.ts
import assert from 'node:assert/strict';
import { execFileSync } from 'node:child_process';
import { mkdirSync, mkdtempSync, renameSync, rmSync, realpathSync, openSync, closeSync } from 'node:fs';
import { dirname, join, relative, resolve } from 'node:path';
import { parseArgs } from 'node:util';
import { archiveFiles, command, fileDigest, isMain, readJson, ROOT, unpackVerified, writeJson } from './lib/artifacts.ts';
import type { Inventory } from './lib/artifacts.ts';

export const INPUTS = ['src/asset-catalog.json', 'src/auxide-samples.json', 'scripts/import-assets.ts'] as const;
const ICONS = ['iron-plate.png', 'iron-gear-wheel.png', 'processing-unit.png', 'programmable-speaker.png'];
const FONTS = ['NotoSans-Regular.ttf', 'NotoSans-Bold.ttf', 'NotoMono-Regular.ttf'];
export type AssetLock = {
  schema: 1; repository: string; tag: string; archive: string; sha256: string; size: number;
  factorio_version: string; inputs: Record<string, string>; files: Inventory;
};
export function checkInputs(lock: AssetLock): void {
  assert.equal(lock.schema, 1, 'Unsupported asset lock schema');
  for (const name of INPUTS) assert.equal(lock.inputs[name], fileDigest(join(ROOT, name)).sha256, `${name} changed; regenerate and publish an asset pack`);
}
export function requiredFiles(directory: string): string[] {
  const { pages } = readJson<{ pages: number }>(join(directory, 'packed/manifest.json'));
  assert(Number.isInteger(pages) && pages >= 1 && pages <= 32, 'Invalid atlas page count');
  return [
    'packed/manifest.json',
    ...Array.from({ length: pages }, (_, i) => `packed/atlas-${i}.png`),
    ...ICONS.map(name => `icons/${name}`), ...FONTS.map(name => `fonts/${name}`),
    ...Object.values(readJson<Record<string, string>>(join(ROOT, 'src/auxide-samples.json'))).map(name => `sound/programmable-speaker/${name}`),
  ].sort();
}
export function verifyInstalled(lock: AssetLock, directory: string): void {
  checkInputs(lock);
  assert.deepEqual(requiredFiles(directory), Object.keys(lock.files).sort(), 'Asset lock does not match required runtime files');
  for (const [name, expected] of Object.entries(lock.files)) assert.deepEqual(fileDigest(join(directory, name)), expected, `Asset differs from the lock: ${name}`);
}
function pack(lockPath: string, tag: string, version: string, output: string): void {
  assert(/^[A-Za-z0-9][A-Za-z0-9._-]+$/.test(tag), 'Invalid asset tag');
  const source = join(ROOT, 'public/factorio');
  const names = requiredFiles(source);
  mkdirSync(output, { recursive: true });
  const archive = join(output, 'factorio-runtime.tar.gz');
  archiveFiles(source, names, archive);
  const lock: AssetLock = {
    schema: 1, repository: 'joshcazalas/website-assets', tag, archive: 'factorio-runtime.tar.gz',
    ...fileDigest(archive), factorio_version: version,
    inputs: Object.fromEntries(INPUTS.map(name => [name, fileDigest(join(ROOT, name)).sha256])),
    files: Object.fromEntries(names.map(name => [name, fileDigest(join(source, name))])),
  };
  writeJson(lockPath, lock); writeJson(join(output, 'assets.lock.json'), lock);
  console.log(`Packed ${names.length} runtime files into ${archive} (${lock.size.toLocaleString()} bytes)`);
}
type AssetRelease = { draft: boolean; prerelease: boolean; immutable: boolean; assets: { id: number; name: string; size: number; digest: string }[] };
function fetchAssets(lock: AssetLock, localArchive?: string): void {
  checkInputs(lock);
  const destination = join(ROOT, 'public/factorio');
  mkdirSync(destination, { recursive: true });
  const work = mkdtempSync(join(ROOT, 'public/.factorio-stage-'));
  try {
    const archive = localArchive ? resolve(localArchive) : join(work, 'assets.tar.gz');
    if (!localArchive) {
      assert.equal(lock.repository, 'joshcazalas/website-assets', 'Unexpected asset repository');
      assert(/^[A-Za-z0-9][A-Za-z0-9._-]+$/.test(lock.tag), 'Invalid asset tag');
      const release = JSON.parse(command('gh', 'api', `repos/${lock.repository}/releases/tags/${lock.tag}`)) as AssetRelease;
      assert(!release.draft && !release.prerelease && release.immutable, 'The asset release must be published and immutable');
      const matches = release.assets.filter(asset => asset.name === lock.archive);
      assert.equal(matches.length, 1, 'Missing asset archive');
      const asset = matches[0];
      assert.equal(asset.size, lock.size); assert.equal(asset.digest, `sha256:${lock.sha256}`);
      const fd = openSync(archive, 'wx');
      try { execFileSync('gh', ['api', '-H', 'Accept: application/octet-stream', `repos/${lock.repository}/releases/assets/${asset.id}`], { stdio: ['ignore', fd, 'inherit'] }); }
      finally { closeSync(fd); }
    }
    assert.deepEqual(fileDigest(archive), { sha256: lock.sha256, size: lock.size }, 'Asset archive checksum or size mismatch');
    const staged = join(work, 'files'); mkdirSync(staged);
    unpackVerified(archive, lock.files, staged); verifyInstalled(lock, staged);
    for (const name of Object.keys(lock.files)) {
      const target = join(destination, name); mkdirSync(dirname(target), { recursive: true });
      const parent = relative(realpathSync(destination), realpathSync(dirname(target)));
      assert(!parent.startsWith('..'), 'Asset destination escapes its root');
      renameSync(join(staged, name), target);
    }
    console.log(`Installed verified assets: ${lock.tag}`);
  } finally { rmSync(work, { recursive: true, force: true }); }
}
if (isMain(import.meta.url)) {
  const { values, positionals } = parseArgs({ allowPositionals: true, options: {
    lock: { type: 'string', default: join(ROOT, 'assets.lock.json') },
    tag: { type: 'string' }, 'game-version': { type: 'string' }, archive: { type: 'string' },
    output: { type: 'string', default: join(ROOT, '.local/asset-packs') },
  }});
  const [action] = positionals;
  if (action === 'pack') {
    assert(values.tag && values['game-version'], 'pack requires --tag and --game-version');
    pack(values.lock, values.tag, values['game-version'], resolve(values.output));
  } else if (action === 'fetch') fetchAssets(readJson<AssetLock>(values.lock), values.archive);
  else if (action === 'verify') verifyInstalled(readJson<AssetLock>(values.lock), join(ROOT, 'public/factorio'));
  else throw new Error('Usage: node scripts/assets.ts pack|fetch|verify [options]');
}

File: scripts/assets.py
#!/usr/bin/env python3
"""Package and verify the exact runtime assets used by the website."""

import argparse
import gzip
import hashlib
import io
import json
import os
from pathlib import Path, PurePosixPath
import re
import subprocess
import tarfile
import tempfile

ROOT = Path(__file__).resolve().parent.parent
INPUTS = ("src/asset-catalog.json", "src/auxide-samples.json", "scripts/import-assets.mjs")
ICONS = ("iron-plate.png", "iron-gear-wheel.png", "processing-unit.png", "programmable-speaker.png")
FONTS = ("NotoSans-Regular.ttf", "NotoSans-Bold.ttf", "NotoMono-Regular.ttf")


def digest(data):
    return hashlib.sha256(data).hexdigest()


def read_json(path):
    return json.loads(Path(path).read_text())


def write_json(path, value):
    Path(path).parent.mkdir(parents=True, exist_ok=True)
    Path(path).write_text(json.dumps(value, indent=2, sort_keys=True) + "\n")


def inventory(directory):
    result = {}
    for path in sorted(Path(directory).rglob("*")):
        if path.is_symlink():
            raise ValueError(f"Symlinks are not allowed: {path}")
        if path.is_file():
            data = path.read_bytes()
            result[path.relative_to(directory).as_posix()] = {"sha256": digest(data), "size": len(data)}
    return result


def archive_files(directory, names, output):
    """Stable archives: sorted paths, fixed modes, no host names or timestamps."""
    with Path(output).open("wb") as raw:
        with gzip.GzipFile(filename="", mode="wb", fileobj=raw, mtime=0) as compressed:
            with tarfile.open(fileobj=compressed, mode="w", format=tarfile.USTAR_FORMAT) as archive:
                for name in sorted(names):
                    path = Path(directory) / name
                    if path.is_symlink() or not path.is_file():
                        raise ValueError(f"Expected a regular file: {name}")
                    data = path.read_bytes()
                    member = tarfile.TarInfo(name)
                    member.size = len(data)
                    member.mode = 0o644
                    archive.addfile(member, io.BytesIO(data))


def unpack_verified(archive_path, expected, destination):
    """Reject traversal, links, duplicates, missing files, and unexpected payloads."""
    seen = set()
    with tarfile.open(archive_path, "r:gz") as archive:
        for member in archive:
            name = member.name
            path = PurePosixPath(name)
            if (not member.isfile() or path.is_absolute() or ".." in path.parts
                    or str(path) != name or name not in expected or name in seen):
                raise ValueError(f"Unexpected archive member: {name}")
            if member.size != expected[name]["size"]:
                raise ValueError(f"Incorrect file size: {name}")
            data = archive.extractfile(member).read()
            if digest(data) != expected[name]["sha256"]:
                raise ValueError(f"Incorrect file checksum: {name}")
            target = Path(destination) / name
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(data)
            seen.add(name)
    if seen != set(expected):
        raise ValueError(f"Missing archive members: {sorted(set(expected) - seen)}")


def check_inputs(lock):
    if lock.get("schema") != 1:
        raise ValueError("Unsupported asset lock schema")
    for name in INPUTS:
        if lock["inputs"].get(name) != digest((ROOT / name).read_bytes()):
            raise ValueError(f"{name} changed; regenerate and publish a new asset pack")


def required_files(directory):
    pages = read_json(Path(directory) / "packed/manifest.json")["pages"]
    if not isinstance(pages, int) or not 1 <= pages <= 32:
        raise ValueError("Invalid atlas page count")
    return sorted([
        "packed/manifest.json",
        *(f"packed/atlas-{i}.png" for i in range(pages)),
        *(f"icons/{name}" for name in ICONS),
        *(f"fonts/{name}" for name in FONTS),
        *(f"sound/programmable-speaker/{name}" for name in read_json(ROOT / "src/auxide-samples.json").values()),
    ])


def verify_installed(lock, directory):
    check_inputs(lock)
    if required_files(directory) != sorted(lock["files"]):
        raise ValueError("Asset lock does not match required runtime files")
    for name, expected in lock["files"].items():
        path = Path(directory) / name
        if path.is_symlink() or not path.is_file():
            raise ValueError(f"Missing regular asset file: {name}")
        data = path.read_bytes()
        if len(data) != expected["size"] or digest(data) != expected["sha256"]:
            raise ValueError(f"Asset differs from the lock: {name}")


def pack(args):
    if not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9._-]+", args.tag):
        raise ValueError("Invalid asset tag")
    source = ROOT / "public/factorio"
    names = required_files(source)
    output = Path(args.output).resolve()
    output.mkdir(parents=True, exist_ok=True)
    archive = output / "factorio-runtime.tar.gz"
    archive_files(source, names, archive)
    files = {name: {"sha256": digest((source / name).read_bytes()), "size": (source / name).stat().st_size} for name in names}
    lock = {
        "schema": 1, "repository": args.repository, "tag": args.tag,
        "archive": archive.name, "sha256": digest(archive.read_bytes()), "size": archive.stat().st_size,
        "factorio_version": args.game_version,
        "inputs": {name: digest((ROOT / name).read_bytes()) for name in INPUTS},
        "files": files,
    }
    write_json(args.lock, lock)
    write_json(output / "assets.lock.json", lock)
    print(f"Packed {len(names)} runtime files into {archive} ({archive.stat().st_size:,} bytes)")


def fetch(args):
    lock = read_json(args.lock)
    check_inputs(lock)
    destination = ROOT / "public/factorio"
    destination.parent.mkdir(parents=True, exist_ok=True)
    # Stage beside the destination so a failed download never damages local assets.
    with tempfile.TemporaryDirectory(prefix=".factorio-stage-", dir=destination.parent) as work:
        work = Path(work)
        archive = Path(args.archive).resolve() if args.archive else work / lock["archive"]
        if not args.archive:
            repo = lock["repository"]
            if repo != "joshcazalas/website-assets":
                raise ValueError("Unexpected asset repository")
            result = subprocess.run(["gh", "api", f"repos/{repo}/releases/tags/{lock['tag']}"], check=True, capture_output=True, text=True)
            release = json.loads(result.stdout)
            if release["draft"] or release["prerelease"] or not release.get("immutable"):
                raise ValueError("The asset release must be published and immutable")
            matches = [asset for asset in release["assets"] if asset["name"] == lock["archive"]]
            if len(matches) != 1 or matches[0]["size"] != lock["size"]:
                raise ValueError("Asset release has no matching archive")
            asset = matches[0]
            if asset.get("digest") != "sha256:" + lock["sha256"]:
                raise ValueError("GitHub's asset digest differs from the pinned digest")
            with archive.open("wb") as output:
                subprocess.run(["gh", "api", "-H", "Accept: application/octet-stream", f"repos/{repo}/releases/assets/{asset['id']}"], stdout=output, check=True)
        if archive.stat().st_size != lock["size"] or digest(archive.read_bytes()) != lock["sha256"]:
            raise ValueError("Asset archive checksum or size mismatch")
        staged = work / "files"
        staged.mkdir()
        unpack_verified(archive, lock["files"], staged)
        verify_installed(lock, staged)
        # Preserve the existing imported originals; builds include only the locked runtime set.
        for name in lock["files"]:
            target = destination / name
            target.parent.mkdir(parents=True, exist_ok=True)
            if target.parent.resolve().is_relative_to(destination.resolve()) is False:
                raise ValueError("Asset destination escapes its root")
            os.replace(staged / name, target)
    print(f"Installed verified assets: {lock['tag']}")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--lock", default=str(ROOT / "assets.lock.json"))
    commands = parser.add_subparsers(dest="command", required=True)
    create = commands.add_parser("pack")
    create.add_argument("--repository", default="joshcazalas/website-assets")
    create.add_argument("--tag", required=True)
    create.add_argument("--game-version", required=True)
    create.add_argument("--output", default=str(ROOT / ".local/asset-packs"))
    create.set_defaults(run=pack)
    download = commands.add_parser("fetch")
    download.add_argument("--archive", help="Use a local pack, with identical checksum verification")
    download.set_defaults(run=fetch)
    verify = commands.add_parser("verify")
    verify.set_defaults(run=lambda args: verify_installed(read_json(args.lock), ROOT / "public/factorio"))
    args = parser.parse_args()
    args.run(args)


if __name__ == "__main__":
    main()
